server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/PreSplit.java

```diff
@@ -98,7 +99,10 @@ public long isReady(long tid, Manager manager) throws Exception {
 
         if (tabletMetadata != null && tabletMetadata.getLocation() != null) {
           // the tablet exists but has a location, lets try again later
-          manager.requestUnassignment(tabletMetadata.getExtent(), tid);
+          // ignore the return value
+          manager.requestTabletUnload(
```

Comment:
Instead of making this thrift call we can set the next operation id directly here.  Experimented with that locally and came up with the following rewrite of `isReady()`.  Also this allows deleting the next opdi and setting the opid in the same mutation.  

```java
    TabletMetadata tablet = manager.getContext().getAmple().readTablet(splitInfo.getOriginal());

    if (tablet == null || tablet.getOperationId() != null
        || !opid.equals(tablet.getNextOperationId())) {
      // The tablet may no longer exists, another operation may have it reserved, or maybe we
      // already reserved and a fault happened. In any case lets proceed, the tablet will be
      // checked in the call() function and it will sort everything out.
      return 0;
    } else if (tablet.getLocation() != null) {
      if (opid.equals(tablet.getNextOperationId())) {
        // we have already set the next opid to ourself, but tablet is still assigned so wait
        // TODO log something
        return 1000;
      }

      // tablet is hosted so need to set next opid to get it unassigned
      try (var tabletsMutator = manager.getContext().getAmple().conditionallyMutateTablets()) {
        tabletsMutator.mutateTablet(splitInfo.getOriginal()).requireAbsentNextOperation()
            .requireSame(tablet, ColumnType.PREV_ROW, ColumnType.LOCATION, ColumnType.OPID)
            .startOperation(opid).submit(tmeta -> opid.equals(tmeta.getNextOperationId()));
        Map<KeyExtent,Ample.ConditionalResult> results = tabletsMutator.process();
        if (results.get(splitInfo.getOriginal()).getStatus() == Status.ACCEPTED) {
          log.debug("{} set next opid {} for split", FateTxId.formatTid(tid), splitInfo.getOriginal());
          // set the next op id, so come back later and see if there is a location
          manager.getEventCoordinator().event(splitInfo.getOriginal(), "Set next opid");
          return 1000;
        } else {
          // failed to set next operation id on tablet, lets try again later and sort things out then
          // TODO log something
          return 1000;
        }
      }
    } else {
      // the following was checked in other parts of the if statement and should be true, defending against changes to the other parts of the code
      Preconditions.checkState(tablet.getOperationId() == null && tablet.getLocation() == null);
      // tablet is not hosted and operation id is not set, so can set it.
      try (var tabletsMutator = manager.getContext().getAmple().conditionallyMutateTablets()) {
        // only mutate tablet if there is no operation id and prev row, location, and OPID_NEXT are
        // the same as what we saw
        var tabletMutator = tabletsMutator.mutateTablet(splitInfo.getOriginal())
            .requireAbsentOperation()
            .requireSame(tablet, ColumnType.PREV_ROW, ColumnType.LOCATION, ColumnType.OPID_NEXT);
        tabletMutator.startOperation(opid);
        if (opid.equals(tablet.getNextOperationId())) {
          // set the operation id and delete the next operation id in the same mutation
          tabletMutator.deleteNextOperation();
        }
        tabletMutator.submit(tmeta -> opid.equals(tmeta.getOperationId()));

        Map<KeyExtent,Ample.ConditionalResult> results = tabletsMutator.process();
        if (results.get(splitInfo.getOriginal()).getStatus() == Status.ACCEPTED) {
          log.debug("{} reserved {} for split", FateTxId.formatTid(tid), splitInfo.getOriginal());
          return 0;
        } else {
          // failed to set operation id on tablet, lets try again later
          // TODO log something
          return 1000;
        }
      }
    }
```

Reply:
> Instead of making this thrift call we can set the next operation id directly here.

This is not calling the Thrift API. The Thrift API methods are in `ManagerClientServiceHandler` which do some validation, then call methods on the `Manager`. This is calling the `Manager` method.

Reply:
> This is not calling the Thrift API. 

I missed that, what is calling the thrift API?

I still think its nicer to set the next opid directly in this code.  Makes it easier to see what is happening.  Can also delete the next opid and set the opid in the same mutation when transitioning.

Reply:
The example code I wrote in the initial comment is not as efficient as the existing code and could be improved.  I was just trying to think through the situations in code.  The existing code starts of with a condition mutation to set the opid and if that fails then it reads the tablet metadata to see whats going on.  That approach is probably still best for efficiency.